### DISCUSSION TITLE
Setting (or not) DYLD_LIBRARY_PATH

### DISCUSSION THREAD

**Discussion by skullydazed** [Homebrew/discussions#1737] [Writing Formulae/Casks] [answered]
I work on [QMK Firmware](https://github.com/qmk/qmk_firmware), a pretty popular firmware for custom mechanical keyboards. We have some users who are starting to run into issues on their M1 macs, but as none of the maintainers have one we have limited ability to work on this ourselves. We recommend users install with rosetta which still works fine, but some users still try to use the AS native version, which has one issue in particular that's proving troublesome- the need to set `DYLD_LIBRARY_PATH`.

Trying to summarize as much as possible, here is my understanding of the situation:

* Users git clone [qmk_firmware](https://github.com/qmk/qmk_firmware) and build firmware for their embedded system (avr-gcc and/or arm-gcc) using the make-based build system
* QMK Firmware provides a CLI (`qmk`, written in python) which provides both user tools and tools used to generate C include files used by the make-based build system.
* This CLI is installed using a brew tap: `brew install qmk/qmk/qmk`
* This means that homebrew's python is used inside a virtualenv, per homebrew standards
* One of the tools this CLI provides requires the [hid](https://pypi.org/project/hid/) module
* `hid` attempts to dynamically load one of several libraries (libhidapi-hidraw.so libhidapi-hidraw.so.0 libhidapi-libusb.so libhidapi-libusb.so.0 libhidapi-iohidmanager.so libhidapi-iohidmanager.so.0 libhidapi.dylib hidapi.dll libhidapi-0.dll)
* This all works fine as long as you have `export DYLD_LIBRARY_PATH=/opt/homebrew/lib` in your shell.
* This breaks when a tool sanitizes your environment
* QMK Firmware attempts to run several `qmk` commands to generate C includes and build rules from inside `make`, but [make seemingly wipes out `DYLD_LIBRARY_PATH`](https://github.com/qmk/qmk_cli/issues/60#issuecomment-867777428).

The problem that we (QMK Firmware) face is what to do here. We can figure out how to ensure that `DYLD_LIBRARY_PATH` is passed in our Makefile, but ISTM* that there are other packages that face similar issues. Since it uses a brew-provided python ISTM that the proper paths could be taught to python so env vars aren't needed. ISTM that [this is a bug](https://github.com/Homebrew/brew/issues/9177#issuecomment-733577675), but I'm not sure how or where to best fix it.

\* Everytime I write ISTM you can mentally replace it with "I'm making a big assumption here and am probably wrong."

**Comment by carlocab** [comment, marked as answer]
`make` doesn't wipe out `DYLD_LIBRARY_PATH`, Apple's System Integrity Protection does. I believe the rule is that it doesn't propagate into subshells, but I don't think that's something we can rely on. Setting it also tends to break things. You really don't want your build to depend on being able to set this variable.

I commented at https://github.com/qmk/qmk_cli/issues/60 to see if there are fixes we can apply in Homebrew.

**Reply by LSChyi** [reply]
I haven't do the patch in Homebrew formula before, so I documented here what I have done and keep the discussion here.

```
$ brew edit python@3.9
# apply the changes for adding the ldflags_nodist of `"-Wl,-rpath,#{HOMEBREW_PREFIX}/lib"`
$ brew reinstall python@3.9 -s
```

After that, importing `hid` is working without the error of missing `libhidapi.dylib`.

**Reply by carlocab** [reply]
I'm working on a PR so we can fix this for everyone without having to patch the formula. Just testing it locally first.

**Comment by carlocab** [comment]
Homebrew/homebrew-core#80052 should fix this.
